src/librustc_builtin_macros/asm.rs

```diff
@@ -275,6 +275,14 @@ fn err_duplicate_option<'a>(p: &mut Parser<'a>, symbol: Symbol, span: Span) {
         String::new(),
         Applicability::MachineApplicable,
     );
+    if p.look_ahead(0, |t| t == &token::Comma) {
+        err.tool_only_span_suggestion(
+            p.token.span,
+            "remove this comma",
+            String::new(),
+            Applicability::MachineApplicable,
+        );
+    }
```

Comment:
@Amanieu, the way I implemented the suggestion with the comma is to create a separate, tool-only suggestion to remove the comma if it's there. Otherwise, the UI makes it look like the comma is part of the option.

The tests will fail, though, because for some reason, the tool-only suggestion is showing up in the UI and in a very bizarre way. Am I doing something wrong? Maybe I should just go back to 13c0caff100cc69a6699fb727a5684fbb1fc10c1.

Here's what it looks like:

```
error: the `att_syntax` option was already provided
  --> rust-issue-73193.rs:10:28
   |
10 |             options(nomem, att_syntax, nostack, att_syntax),
   |                            ^^^^^^^^^^
   |
help: remove this option
   |
10 |             options(nomem, , nostack, att_syntax),
   |                           --
```

Reply:
I'm a bit out of my depth here, so I will delegate to @estebank who is more familiar with the diagnostic infrastructure.

Reply:
I'm thinking about just going back to 13c0caf, and have it highlight the option and the comma, and then I can revise it in a later pull request. What do you think?

Reply:
I believe that here you can use

```rust
let comma_sp = p.look_ahead(1, |t| t.span);
p.token.span.to(comma_sp)

```

to point at `att_syntax,`, although ideally, you would do

```
let next_ident = p.look_ahead(2, |t| t.span);
p.token.span.until(next_ident)
```

to point at `att_syntax, `

Reply:
@estebank In 13c0caf, I was able to point at `att_syntax,`, but I would rather just point at `att_syntax` in the UI, but have the tooling see `att_syntax,`.

So, my goal for the UI is:

```
error: the `att_syntax` option was already provided
  --> rust-issue-73193.rs:10:28
   |
10 |             options(nomem, att_syntax, nostack),
   |                            ^^^^^^^^^^
```

And my goal for the tooling is the suggestion:

```diff
- options(nomem, att_syntax, nostack)
+ options(nomem, , nostack)
```

Reply:
I just realized that [the issue](https://github.com/rust-lang/rust/pull/73227#discussion_r441102881) with the current version (c12fad1) is that:

1. the help message is for some reason shown as a separate message, and not as a label
2. and the help message shows the code with the suggestion applied

Why is that? Is the second issue a bug?

Reply:
@estebank could you take a look at this? Thanks!

Reply:
We have a `tool_only_suggestion` that will _not_ be shown in the cli output, but the suggestion will still appear in VSCode and be applicable by `rustfix`. That's what I normally use for a case like this.

The display of structured suggestions are _normally_ as a separate window with the change applied. There's an optimization to show it as a label with the suggested code as part of the label, but there are a bunch of rules on when it doesn't get displayed that way (length of label, if there are more than one suggestion, devel selectable with `span_suggestion_verbose` or `span_suggestion_short`).

Reply:
Hmm. So what do you recommend I do? Because the help message is confusing for the users:

```
error: the `att_syntax` option was already provided
  --> rust-issue-73193.rs:10:28
   |
10 |             options(nomem, att_syntax, nostack, att_syntax),
   |                            ^^^^^^^^^^
   |
help: remove this option
   |
10 |             options(nomem, , nostack, att_syntax),
   |                           --
```

(The help message only appears separately and it only highlights the wrong part when the tool-only suggestion for the comma is there. I tested it with a user-visible suggestion too, and it messed up the help message as well.)

Is it not possible to have multiple suggestions on a diagnostic (or am I doing it wrong)?

Thanks for your help!

Reply:
there's a `multipart_suggestion` method that takes a `Vec<(Span, String)>` that can be used, but that shouldn't be necessary for this case, I think. What I would do is have a `span_label` pointing at `att_syntax` saying "option already provide it, remove it" and a hidden structured suggestion with a span between the start of `att_syntax` to the start of `nostack` for the removal. That way in VSCode you'd see

```
option already provided, remove it
remove the option
```

which is not perfect, but serviceable.

Reply:
Okay, I’ll try that. Thanks for helping me figure this out!

Reply:
@Amanieu What do you think of this output?

```
error: the `att_syntax` option was already provided
  --> rust-issue-73193.rs:10:28
   |
10 |             options(nomem, att_syntax, nostack, att_syntax),
   |                            ^^^^^^^^^^ this option was already provided
```